File: backend/src/mhglauncher/services/notes.py
from __future__ import annotations

import json

from mhglauncher.database import Database
from mhglauncher.models import DailyNote, GameRole
from mhglauncher.providers.base import Provider


class NoteService:
    def __init__(self, database: Database, provider: Provider) -> None:
        self.database = database
        self.provider = provider

    async def refresh(
        self,
        credential: str,
        role: GameRole,
        xrpc_challenge: str = "",
    ) -> DailyNote:
        note = await self.provider.get_daily_note(
            credential,
            role,
            xrpc_challenge,
        )
        await self.database.execute(
            """
            INSERT INTO notes(uid, payload, refreshed_at) VALUES(?, ?, ?)
            ON CONFLICT(uid) DO UPDATE SET payload=excluded.payload,
            refreshed_at=excluded.refreshed_at
            """,
            (role.uid, note.model_dump_json(), note.refreshed_at.isoformat()),
        )
        return note

    async def verify(
        self,
        credential: str,
        challenge: str,
        validate: str,
    ) -> str:
        return await self.provider.verify_note_challenge(
            credential,
            challenge,
            validate,
        )

    async def get(self, uid: str) -> DailyNote | None:
        row = await self.database.fetch_one("SELECT payload FROM notes WHERE uid=?", (uid,))
        if row is None:
            return None
        return DailyNote.model_validate(json.loads(row["payload"]))

File: backend/src/services/notes.ts
import type { DailyNote, GameRole } from "../core/models";
import type { Store } from "../core/database";
import type { Provider } from "../providers/provider";

export class NoteService {
  constructor(private readonly store: Store, private readonly provider: Provider) {}

  async refresh(credential: string, role: GameRole, challenge = ""): Promise<DailyNote> {
    const note = await this.provider.getDailyNote(credential, role, challenge);
    this.store.db.prepare(`INSERT INTO notes(uid,payload,refreshed_at) VALUES(?,?,?)
      ON CONFLICT(uid) DO UPDATE SET payload=excluded.payload,refreshed_at=excluded.refreshed_at`)
      .run(role.uid, JSON.stringify(note), note.refreshed_at);
    return note;
  }

  verify(credential: string, challenge: string, validate: string): Promise<string> {
    return this.provider.verifyNoteChallenge(credential, challenge, validate);
  }

  get(uid: string): DailyNote | null {
    const row = this.store.one("SELECT payload FROM notes WHERE uid=?", uid);
    return row ? JSON.parse(String(row.payload)) as DailyNote : null;
  }
}
